### PR TITLE
Add a quirk for COD detection

### PR DESCRIPTION
**Pull request by superm1**
Insyde has added a bit into the ACPI UEFI table that can be utilized to detect if they have fixed the COD bug.

This means that COD can be turned on by default for X86 for all vendors that support it, and if Insyde is encountered whenever that bit is active.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+			    sizeof(guid),
+			    0x0, /* dst */
+			    buf,
+			    bufsz,
+			    0x24, /* src */
+			    sizeof(guid),
+			    error))
+		return NULL;
+	guidstr = fwupd_guid_to_string(&guid, FWUPD_GUID_FLAG_MIXED_ENDIAN);
+	g_debug("GUID: %s", guidstr);
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:98]: I'd much prefer these g_debug's to be in `firmware->export` tbh

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+#include <fwupdplugin.h>
+
+#include <string.h>
+
+#include "fu-acpi-uefi.h"
+
+struct _FuAcpiUefi {
+	GObject parent_instance;
+	gboolean is_insyde;
+	gboolean working_cod;
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:18]: both in `firmware->export` pls

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+		if (!fu_memcpy_safe((guint8 *)&data_offset,
+				    sizeof(data_offset),
+				    0x0, /* dst */
+				    buf,
+				    bufsz,
+				    0x34, /* src */
+				    sizeof(data_offset),
+				    error))
+			return NULL;
+		if (!fu_memcpy_safe((guint8 *)&working_cod,
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:115]: both use `fu_memread_uint32_safe` perhaps?

File: plugins/uefi-capsule/fu-uefi-capsule-plugin.c [hunk 4]
```diff
@@ -874,18 +876,33 @@ fu_uefi_capsule_plugin_check_cod_support(FuPlugin *plugin, GError **error)
...
+		return TRUE;
+	}
+
+	if (!fu_acpi_uefi_working_cod(acpi_uefi)) {
 		g_set_error_literal(error,
 				    G_IO_ERROR,
 				    G_IO_ERROR_NOT_SUPPORTED,
-				    "Capsule-on-Disk is supported, but non-aarch64 platforms "
-				    "require the 'uefi-allow-cod' quirk for it to be used");
+				    "Capsule-on-Disk may have a firmware bug");
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-uefi-capsule-plugin.c:903]: should we leave uefi-allow-cod or remove it? I guess it might be useful for other vendors.

Comment (superm1) [reply plugins/uefi-capsule/fu-uefi-capsule-plugin.c:903]: I would rather remove it unless we find another bug.

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+			    bufsz,
+			    0x24, /* src */
+			    sizeof(guid),
+			    error))
+		return FALSE;
+	self->guid = fwupd_guid_to_string(&guid, FWUPD_GUID_FLAG_MIXED_ENDIAN);
+	self->is_insyde = (g_strcmp0(self->guid, FU_EFI_INSYDE_GUID) == 0);
+
+	if (self->is_insyde && length >= UEFI_HEADER_SIZE + UEFI_INSYDE_WORKING_COD) {
+		guint32 data_offset;
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:123]: add `= 0x0` to make coverity happy.

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+#include <fwupdplugin.h>
+
+#include <string.h>
+
+#include "fu-acpi-uefi.h"
+
+#define FU_EFI_INSYDE_GUID "9d4bf935-a674-4710-ba02-bf0aa1758c7b"
+
+#define UEFI_HEADER_SIZE	0x36
+#define UEFI_INSYDE_WORKING_COD 0x201
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:18]: Maybe UEFI_TABLE_OFFSET_INSYDE_COD_STATUS or something?

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+struct _FuAcpiUefiTable {
+	FuFirmware parent_instance;
+	gboolean is_insyde;
+	guint32 cod_working;
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:23]: `insyde_cod_status` perhaps?

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+			 GBytes *fw,
+			 gsize offset,
+			 FwupdInstallFlags flags,
+			 GError **error)
+{
+	FuAcpiUefiTable *self = FU_ACPI_UEFI_TABLE(firmware);
+	gchar creator_id[5] = {'\0'};
+	gchar oem_table_id[9] = {'\0'};
+	gchar signature[5] = {'\0'};
+	guint32 length;
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:53]: `= 0x0` pls

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+	if (!fu_memcpy_safe((guint8 *)oem_table_id,
+			    sizeof(oem_table_id),
+			    0x0, /* dst */
+			    buf,
+			    bufsz,
+			    0x10, /* src */
+			    sizeof(oem_table_id) - 1,
+			    error))
+		return FALSE;
+	self->oem_id = g_strdup(oem_table_id);
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:97]: `fu_strsafe` would make me feel warm and fuzzy

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+	if (!fu_memcpy_safe((guint8 *)creator_id,
+			    sizeof(creator_id),
+			    0x0, /* dst */
+			    buf,
+			    bufsz,
+			    0x1c, /* src */
+			    sizeof(creator_id) - 1,
+			    error))
+		return FALSE;
+	self->creator_id = g_strdup(creator_id);
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:107]: `fu_strsafe` would make me feel warm and fuzzy

**Review by hughsie — DISMISSED** [review]
One microscopic nit then LGTM. Thanks!

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+	if (!fu_memcpy_safe((guint8 *)oem_table_id,
+			    sizeof(oem_table_id),
+			    0x0, /* dst */
+			    buf,
+			    bufsz,
+			    0x10, /* src */
+			    sizeof(oem_table_id) - 1,
+			    error))
+		return FALSE;
+	self->oem_id = fu_strsafe(oem_table_id);
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:97]: needs the `sizeof(oem_table_id)`

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+static gboolean
+fu_acpi_uefi_parse_insyde(FuAcpiUefiTable *self, const guint8 *buf, gsize bufsz, GError **error)
+{
+	struct FuAcpiInsydeQuirkSection quirk_section;
+	guint32 data_offset = 0;
+	guint8 *data_start;
+
+	data_start = (guint8 *)g_strstr_len((gchar *)buf, bufsz, "$QUIRK");
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:54]: `fu_memmem_safe()` is going to be a lot quicker -- and probably more safe too :)

File: plugins/uefi-capsule/fu-acpi-uefi.c [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+	gboolean is_insyde;
+	gchar *creator_id;
+	gchar *oem_id;
+	gchar *guid;
+};
+
+struct FuAcpiInsydeQuirkSection {
+	gchar signature[6];
+	guint32 size;
+	guint32 flags;
```
Comment (hughsie) [inline plugins/uefi-capsule/fu-acpi-uefi.c:31]: maybe add a `\* le *\` comment for our futureselves? Also, I think you need to sprinkle some GUINT32_FROM_LE's around.

**Comment by hughsie** [comment]
@superm1 do we want this in 1.8.11?

**Comment by superm1** [comment]
> @superm1 do we want this in 1.8.11?

It needs to be tested first. If testing makes it, yes.

File: plugins/uefi-capsule/fu-acpi-uefi.c (hunk 13 above)
Comment (hughsie) [reply plugins/uefi-capsule/fu-acpi-uefi.c]: I added this.

File: plugins/uefi-capsule/fu-acpi-uefi.c (hunk 12 above)
Comment (hughsie) [reply plugins/uefi-capsule/fu-acpi-uefi.c:54]: I did this.

**Comment by hughsie** [comment]
> It needs to be tested first. If testing makes it, yes.

With a little tweak, this seems to work fine with real data from Insyde.

**Pull request merged**